### ISSUE TITLE
Functional Application

### ISSUE THREAD

**Issue by confused-Techie** [confused-Techie/atom-backend#33]
Now that it was possible to test with the full data set of packages, it seems the current design of the backend **must** change.

Initially on datasets of 100 packages or less, things seemed rather functional. But once within Cloud File Storage, the application was tested against the actual value of about 11,000 unique packages. At this point the backend became unreasonably slow. A request for all packages sorted and pruned would take <300 seconds, at which point the server hosting the backend, cut off the response due to assumed inoperability. This happens because the backend was trying to read each file one by one. 

At this point it was obvious a file store wouldn't work, and there were hopes of moving the full JSON objects into an online Database.  
This functionality was initially tested alongside full file store functionality, but even then, interacting with the DB as if it were a file system, while provided improvement, to 40 seconds reads of all packages, with cached reads about 8 ms. This still wasn't enough.

At this point it should seem obvious that huge portions of the codebase will have to be rewritten to work within a database, and abandon all file store functionality. 

This goes out as an update to any that are contributing on the project, that while the efforts so far have been fantastic, one huge push is needed to get it worked to our expectations.

**Comment by Digitalone1** [comment]
I have a website working with an internal sql database. Previoulsy mySql, then migrated to MariaDB. 

Don't know if I can be of any help since the implementation I use is in PHP.

**Comment by confused-Techie** [comment]
@Digitalone1 if you'd like to link it I'd be happy to take a look at it. Currently I've started to implement it similarly to how `resources.js`. You can look at `database.js` to see some of the functions that are there. If familiar with the syntax for Postgres feel free to add it to those functions where its missing, as I'll be adding more soon. But thats good to know you're familiar with it. Since the syntax is new to me

**Comment by Digitalone1** [comment]
Never used PostgreSQL, but the queries should be in SQL languages like MariaDB, so I could do something with it.

**Comment by Digitalone1** [comment]
Anyway, since we are handling with the database, could you report how the data is structured?
How many tables are stored, how data is organized, how records are indexed and so on...

**Comment by confused-Techie** [comment]
@Digitalone1 currently the data is stored rather generically, but hopefully with time it will be further expanded to properly suit the data. I'll do a write up on how its stored within the repo, and ping you once its there.

**Comment by Digitalone1** [comment]
Well, this is not optimal. Since you use a database, first you have to organize the tables, then the relations between them.

This helps also to speed up the reads and writes. Choosing the proper primary key for packages makes it efficient to search between them. 

This has to be done otherwise also with the database you can have very high load times.

**Comment by confused-Techie** [comment]
@Digitalone1 I understand the concern and am aware its not optimal. Although postgres does index jsonb datatypes, which is what its stored as. So while we will want to migrate in the future, we need to do what we can to have a functional application sooner rather than later. 

But for example currently if I wanted to query every single version of every single package like so

```sql
SELECT pointer, versions FROM packages, jsonb_object_keys(data->'versions') AS versions
```

That currently returns 69959 rows, and takes 1.530 seconds. Which should be fast enough for our current concerns, while of course it can be faster.

But for the meantime, I finally got around to creating a quick docs of the database to help out any contributors looking at it. 
The [database docs are within './docs/'](https://github.com/confused-Techie/atom-community-server-backend-JS/blob/main/docs/database.md)

**Comment by Digitalone1** [comment]
That is a read of all records. It's generally fast. The problem is when you add the WHERE clause and start to select records with conditions.

Did you test it? For example selecting packages of a single user. Or a simple search by name with the ORDER BY clause to show packages with more stars.

I hope it's quite fast also in these cases.

**Comment by confused-Techie** [comment]
@Digitalone1 You do have a point in decreased performance with more complex queries, heres some of the testing that I've done so far, please feel free to let me know if you see anything thats cause for concern going forward, since I'd like to get it up and then work on normalization, but if we should normalize first, then get started that's just what we have to do.

```sql
SELECT pointer FROM pointers WHERE name='teletype';
```

Time: 00:00:00.152

Now another example that I've currently added to the codebase. Creating an array of results, of all packages, ordered by the `downloads` count. Additionally its limited to 30 items, with no offset. Representing the default parameters for `GET /api/packages/`.

```sql
SELECT ARRAY (SELECT data FROM packages ORDER BY data->'downloads' DESC LIMIT 30 OFFSET 0);
```

Time: 00:00:02.064

So while not the fastest, it seems acceptable to me for functionality, especially considering that previously with no database this same exact request would take <300 seconds. Eventually timing out. But what do you think?

**Comment by confused-Techie** [comment]
@Digitalone1 Also on the note of searching, since you're more familiar I was hoping to keep the use of the custom LCS implementation I made, and while I could retrieve all rows of the pointer table and use that for some relatively fine performance, I was curious how difficult or even possible would it be to implement that on the DB itself. Or do you think performance will be markedly better using included search capabilities

**Comment by Digitalone1** [comment]
Two seconds does not seem so good. Take in consideration that the server may be loaded with more requests at the same time, the response might be longer.

Can you make also a search by name? Ordering also by download count. 

I'll tell you what it's happening in this case. I suppose you are using a default primary key which should be something such as an integer incremented every time a new record is stored.

If you want to pick a specific package the system has to scan all the records, even if it already found that package because it does not know that the package is unique.

If you specify a key on the package name, the search will be faster because the DB creates an index on that column to access elements quite instantly (I assume it uses a sort of hash table).

So I think it's important to establish which is the relation between the packages and the other entities involved in the process.

For example, the package names must be unique? Let's make the name as primary key. The name can be duplicated, but a single user must have different package names? The pair of package name and username can be be set as primary key.

Then there's also the question about redundancy. If the package contains the user data and a user stores 10 packages, you have data repeated 10 times in the database which is a waste of space.

Packages and users are separate entities, they should be in different tables and linked by foreign keys.

If you tell me which are the entities and which relations exist between them, I can write a list of tables to store on the DB.

I will also take a look to the full package in the documentation when have time to make an idea. Anyway the first thing to take in consideration is how packages are uniquely identified.

**Comment by Digitalone1** [comment]
> @Digitalone1 Also on the note of searching, since you're more familiar I was hoping to keep the use of the custom LCS implementation I made, and while I could retrieve all rows of the pointer table and use that for some relatively fine performance, I was curious how difficult or even possible would it be to implement that on the DB itself. Or do you think performance will be markedly better using included search capabilities

I don't remembered that part, can you link it?

**Comment by confused-Techie** [comment]
@Digitalone1 Appreciate all the information! And that is a good point about multiple requests slowing service down.

As for the keys I'm still using those UUID's generated during package creation. Since the User, nor package Name are effective keys, since the name can change, and the user can be migrated. But an interesting thing is that once a name has been used, it will always point to the same package, even when that packages name changes. Thats why I originally made the package pointer system. 

In this case now there is a table with package names, and the package UUID associated. There can be several rows that point to the same UUID, since the name can change, but the previous one still points to the same package. 

Then another table uses these UUIDs as the primary key. So that if you need to get a package by name you retrieve the UUID like so 

```sql
SELECT pointer FROM pointers WHERE name=${name}
```

Now you have its primary key for the `packages` table. So then based off what you're saying hopefully that means for searching operations it will know the package is unique when referred to by the UUID and it won't have to scan all records?

And if you are taking a look at this, I don't have a table yet for the users, if thats something you'd be willing to take a look at as well? Since tracking what user creates a package doesn't seem as important, since the ownership of a package is determined based on the access level of that package from the user on GitHub. So really it'd be more important to establish a relationship between the users, and the packages they have stared. 

Lastly sorry if I forgot to ping this to you, but heres a really short write up on the way the db is structured if it helps. 
https://github.com/confused-Techie/atom-community-server-backend-JS/blob/main/docs/database.md

**Comment by Digitalone1** [comment]
> Since the User, nor package Name are effective keys, since the name can change, and the user can be migrated. But an interesting thing is that once a name has been used, it will always point to the same package, even when that packages name changes. Thats why I originally made the package pointer system.

The fact that they change could not be an issue. I have a username and I can change it, but will I be able to set it the same as another existing user? I hope not. The package I created can be changed, but can I have two packages with the same name? I hope not.

> So that if you need to get a package by name you retrieve the UUID like so
> 
> ```sql
> SELECT pointer FROM pointers WHERE name=${name}
> ```
> 
> Now you have its primary key for the `packages` table. So then based off what you're saying hopefully that means for searching operations it will know the package is unique when referred to by the UUID and it won't have to scan all records?

Yes, you won't scan all records if the search is made by UUID, the access is quite direct.

> ```sql
> SELECT * FROM packages WHERE pointer = ${uuid}
> ```

This query is optimized to be faster.

> And if you are taking a look at this, I don't have a table yet for the users, if thats something you'd be willing to take a look at as well? Since tracking what user creates a package doesn't seem as important, since the ownership of a package is determined based on the access level of that package from the user on GitHub. So really it'd be more important to establish a relationship between the users, and the packages they have stared.

I didn't understand this part. I have to admit I didn't fully get what this server backend is managing. Isn't it for packages that can be installed in atom/pulsar as extension? If I create a package, how can I lose it's ownership? This is not even happening on Github. If I make a repository, I have it's ownership, unless I delete it.

**Comment by confused-Techie** [comment]
@Digitalone1 Yeah no worries on duplicate packages or user names, since once you create a package that name is taken, really permanently, or at least thats how it was setup on the original Atom Repository. Even if you change its name, that original one you used always points to the package.

But what I was trying to say on the users. You mentioned the relation to users and the packages. I was more trying to say that the relation of creating packages is not very important. Since the backend doesn't have to store that info. Its more important for the backend to store what a user stars.

But yes the backend is managing the packages that can be installed for Atom/Pulsar. But in short what happens when you create a package here.

You query the backend with your repo saying I want to make this package. Providing the `owner/repo` from there the backend ensures that the name of your repo isn't taken. Then it'll use a GitHub API to find out what repos your GitHub account owns, and if you own the repo specified here then it'll start the publishing process. So thats to say we don't need to do anything on the database end to determine ownership, like having users linked to packages. And theres no endpoint that asks for the packages a user owns at all. 

As for the query you made looking for a pointer that is what happens when you ask for a specific package. For example its already in use here https://github.com/confused-Techie/atom-community-server-backend-JS/blob/d22ba218d5300b12011926457a156664ec2f2552/src/database.js#L43-L46

**Comment by Digitalone1** [comment]
When I have time I'll do a pseudo SQL code with CREATE TABLE.

If you say the package is not related to the user, then the current index with UUID can stay.

For performance it would be better if the package names were unique. But if they can be duplicated, the UUID index is good, no need to do like atom having the first name pointing to the package. The UUID is enough.

But since you say it's important to count the stars given by users, we still need a different table for users.

And, as far as I remember, packages can also have different versions, so we should also have a separate table listing the versions.

@confused-Techie I will give a look at the full package object anyway, but if there are other entities involved, list them here, so I won't forget.

**Comment by confused-Techie** [comment]
I appreciate you taking a look at it. But yes the users do need their own table, and there should be a user object example in the docs, or within the `./data/users.json` file can be a good reference. Since thats something that I'm not sure what the best methodology is for creating that table. Otherwise yeah with the packages there is the basic data that it needs, like the readme, metadata, created_at, updated_at and so on. But the important bits would be things like the star_gazers, which originally was an array of the user names of who has stared a package, plus for each user they also had a list of all packages they have starred. Hopefully this redundancy can be removed. Then like you said the versions of each package. 

As for constructing the table, I'd suggest looking at the `server_package_object` since it contains additional data that the full package does not.

**Comment by Digitalone1** [comment]
So this is a first sketch

 ```sql
CREATE TABLE Packages (
    ID INTEGER AUTO_INCREMENT NOT NULL,
    Name VARCHAR(256) NOT NULL,
    Description TEXT,
    Repository JSON,
    Keywords TEXT,
    Download UNSIGNED INT NOT NULL,
    PRIMARY KEY (ID)
);

CREATE TABLE PackageVersions (
    ID INTEGER AUTO_INCREMENT NOT NULL,
    PackageID INTEGER NOT NULL,
    SemVer VARCHAR(50) NOT NULL,
    Engine JSON,
    License VARCHAR(50),
    PRIMARY KEY (ID),
    FOREIGN KEY (PackageID) REFERENCES Packages(ID)
);

CREATE TABLE Dependencies (
    PackageVersionID INTEGER NOT NULL,
    PackageID INTEGER NOT NULL,
    Relation VARCHAR(256) NOT NULL
    PRIMARY KEY (PackageVersionID, PackageID),
    FOREIGN KEY (PackageVersionID) REFERENCES PackageVersions(ID)
    FOREIGN KEY (PackageID) REFERENCES Packages(ID)
);

CREATE TABLE Users (
    ID INTEGER AUTO_INCREMENT NOT NULL,
    PulsarToken VARCHAR(256) NOT NULL,
    GithubToken VARCHAR(256) NOT NULL,
    PRIMARY KEY (ID),
);

CREATE TABLE Stars (
    PackageID INTEGER NOT NULL,
    UserID INTEGER NOT NULL,
    PRIMARY KEY (PackageID, UserID),
    FOREIGN KEY (PackageID) REFERENCES Packages(ID)
    FOREIGN KEY (UserID) REFERENCES Users(ID)
); 
```

Note that this is in MySQL/MariaBD, the syntax can be slightly different in PostgreSQL, but the concept should be identical.

The `Packages` table represents the entity of packages which can have one or more versions listed in the `PackageVersions`. Now the foreign keys are important because they ensure the integrity of the data stored into the DB.

I can't store in `PackageVersions` a version related to an unexisting package, which means when I fill the `PackageVersions` table, the `PackageID` has to be one of the `ID` present in `Packages`. If I try to write an ID not present in `Packages`, it's an error, the engine does not allow the operation.

At the same time, if I want to delete a package, I can't do it if there are one or more related versions. If there are some versions, I have to pick the `ID` from `Packages`, then go to `PackageVersions` and delete all records with `PackageVersions.PackageID` = `Package.ID`. At last you can delete the `ID` from `Packages`.

There are other things to implement anyway, such as the `configSchema` which I don't know what it means. This is incomplete, but it's done as a first reference.

@confused-Techie if you have doubts, feel free to ask. If I have time, I'll try to do a SELECT query for star count.

**Comment by Digitalone1** [comment]
For `Stars` and `Dependencies` I made a **many to many** relationship. Please take a look [here](https://dzone.com/articles/how-to-handle-a-many-to-many-relationship-in-datab).

A version (of a package) can have more packages as dependencies and a package can be a dependency of more versions (the `Relation` column can be for example `^0.4.1`).

A user can prefer more packages (giving a star to them) and a package can have more preferences (stars from users).

For getting packages and their stars:

```
SELECT Packages.Name, COUNT(UserID) AS StarCount,
FROM Packages, Stars,
WHERE Packages.ID = Stars.PackageID
GROUP BY Stars.PackageID
ORDER BY StarCount DESC
```

I do not guarantee this works. I would have to test it (and maybe in PostgreSQL the query should be different).

**Comment by confused-Techie** [comment]
@Digitalone1 Sorry for a late reply, but looking at this looks fantastic, definitely a bit more complex than what's currently setup. I really do like the idea of storing stars within their own table, instead of duplicating the data across two separate tables. And removes the concerns that were there with JSON files, about incomplete removal or addition to stars. 

I'll have to see about adding this to it, and making sure everything's compatible, which it all looks rather familiar.

**Comment by Digitalone1** [comment]
@confused-Techie when I have time I'll try to write the previous statements in PostgreSQL, so you could try to execute them to build the tables directly. I'm not familiar with it, I have to look at the documentation. Maybe it's similar, but I'm pretty sure some types have to be rewritten along with the syntax related to the primary/foreign keys..

**Comment by Digitalone1** [comment]
@confused-Techie I made some researches. The relational model above is good, but it requires to convert our JavaScript data in DB data. For example we have to extract name, description repository data, keywords and download count to make an insert query in order store a new package.

With JSON we can pass it directly (entirely or a portion). It's more straightforward. But the good of the relational model is that it guarantees data integrity using foreign keys, which is crucial for our scopes, especially when we are dealing with package dependencies and user stars.

So my proposal is to keep the tables and their keys as they are defined above, but the rest of the data is stored ad JSON in a specific column. Or rather, in `jsonb` which gives some advantages.

**Comment by Digitalone1** [comment]
Here are the syntax in PostgreSQL

```sql
CREATE TABLE packages (
    id SERIAL PRIMARY KEY,
    data JSONB NOT NULL
);

CREATE TABLE packageVersions (
    id SERIAL PRIMARY KEY,
    packageId SERIAL REFERENCES packages(id), /* foreign packages key */
    data JSONB NOT NULL
);

CREATE TABLE Dependencies (
    packageVersionId SERIAL REFERENCES packageVersions(id), /* foreign package versions key */,
    packageId SERIAL REFERENCES packages(id),               /* foreign packages key */,
    relation VARCHAR(32) NOT NULL,                          /*i.e. "^0.4.1"*/
    PRIMARY KEY (packageVersionId, packageId)
);

CREATE TABLE users (
    id SERIAL PRIMARY KEY,
    data JSONB NOT NULL
);

CREATE TABLE Stars (
    packageId SERIAL REFERENCES packages(id), /* foreign packages key */,
    userId SERIAL REFERENCES users(id),       /* foreign users key */,
    PRIMARY KEY (packageId, userId)
)
```

@confused-Techie how do you want to proceed?

**Comment by confused-Techie** [comment]
This is fantastic thank you @Digitalone1. I do like the idea of pushing some of the data into the jsonb datatype, and leaving as is. Additionally looking at it now it seems that the `Dependencies` may not be needed since I don't see any logic that interacts with them at all. And that data is just returned as JSON for the packages. But will absolutely implement the `users` and `Stars` right away as those will be the last most needed tables. Otherwise I will need to find a way to programmatically populate the `packageVersions` table. But otherwise this is fantastic work

**Comment by confused-Techie** [comment]
@Digitalone1 I've gone ahead and created nearly all tables based off the provided recommendations, except for leaving the packages table as is, since the data has already been migrated.

I've gone ahead and included the [create_scripts](https://github.com/confused-Techie/atom-community-server-backend-JS/tree/main/scripts/database) in the repo.

Thanks for the recommendations, and suggestions.

**Comment by Digitalone1** [comment]
Did you migrate the data?

It's good having the uuid, but you have to remove the `versions` object from `package_object_full` stored into the `packages` and move it inside the `package_versions`.

```sql
CREATE TABLE packages (
    uuid UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
    data JSONB NOT NULL
);

CREATE TABLE packageVersions (
    id SERIAL PRIMARY KEY,
    packageId UUID REFERENCES packages(id), /* foreign packages key */
    data JSONB NOT NULL
);
```

For migration, I suppose you have all the full packages collected into an array. The script shoud be something like this

```js
for (let package of array_of_package_object_full) {
  let versions = package.versions; /* save versions */
  
  delete package.versions; /* remove versions from package */
  
  const json_package = JSON.stringify(package);
  
  const uuid = await sql_storage`
    INSERT INTO packages (data)
    VALUES (${json_package})  
    RETURNING uuid
  `; /* It should return the uuid of the new row */
    
  /* Now let's store versions */
  
  for (let ver in versions) {
    const json_version = JSON.stringify(version[ver]);
    
    await sql_storage`
      INSERT INTO package_versions (packageId,data)
      VALUES (${uuid},${json_package})
    `; /* store version object with relative foreign key */
  }
}
```

**Comment by Digitalone1** [comment]
> EDIT 2: Also I've noticed you're change from the `users` table to containing just the jsonb object, with the fact that as its currently implemented we only would need the `atom_token`, `github_token` and `created_at` values tracked, maybe we could use just that to have a fully normalized table. Additionally this may actually be helpful since auth is only with a provided token, and having it normalized could allow us to check all possible `atom_token`s rather easily.

No problem, you can create the needed columns. I suppose the tokens are fixed length, so they would be `CHAR(num)`.

Regarding `created_at`, postgresql tutorial suggests `TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP`, so at row creation you wouldn't even specify a value, it's automatically set the current timestamp.

Anyway I'm a bit concerned about the data integrity because a spammer could change their name and register as a new user to leave more votes? Could that be possible? How can we identify a user uniquely even if they changes name?

**Comment by confused-Techie** [comment]
Ideally when a user changes their name the normal way, the `id` would stay the same so they couldn't cast additional votes. But if they were to create a totally new account then there is zero logic to prevent them from casting additional votes. 

One hump to that attempt of spam would be that an account is linked to GitHub, and likely will share the user name there, so that we don't need any logic to change the user name, it would just be changed on GitHub and then we need an updated token relating to that new name, since having the old name would cause current ownership logic to fail. It's not the most robust of plans, but is how it'll currently happen, since admittedly it wasn't one of my top priorities. 

Beyond that we aren't doing anything to stop it, and personally I'm not sure thats within our scope to try and stop spam stars to that extent. Really I feel our priority would be to stop spam packages from entering the system, but I could be completely wrong on that.

Would you have any ideas on how we could go ahead and attempt to prevent spam stars?

Also for your comments on the `created_at` that does sound so much more convenient. Before you mentioned it, I did create that column with the `bigint` datatype, but will see about changing that to `TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP` since that sounds very helpful

**Comment by mauricioszabo** [comment]
So, my 2 cents (didn't look at the full discussion, sorry) is that we could tie an user to a group of repos, like:

`users` -> `git_repositories` and `git_repositories` could have for example the kind (GitHub,  Gitlab, Gitea at least) maybe URL (for self-hosted instead of cloud repos) and then we could integrate with respective git providers to have tokens (that way we won't have Auth information on our side, which is a big plus for me).

As of preventing SPAM, there are some tutorials on using SpamAssassin from other projects, so maybe we can check for packages with READMEs and other docs that have high points and alert something, so we can approve or reject them?

**Comment by Digitalone1** [comment]
> `users` -> `git_repositories` and `git_repositories` could have for example the kind (GitHub, Gitlab, Gitea at least) maybe URL (for self-hosted instead of cloud repos) and then we could integrate with respective git providers to have tokens (that way we won't have Auth information on our side, which is a big plus for me).

That sounds good. With could also use the tokens as primary key. 
 
> As of preventing SPAM, there are some tutorials on using SpamAssassin from other projects, so maybe we can check for packages with READMEs and other docs that have high points and alert something, so we can approve or reject them?

It's doable, but for the moment I suggest to focus on the database structure since it's configuration can prevent issues in the future. Then maybe we can think about additional methods to prevent spam.

**Comment by Digitalone1** [comment]
@confused-Techie did you read my previous post about the data migration? Tell me if there's something wrong in the script.

Please, when you finish, update the database structure into `docs/database.md`.

**Comment by confused-Techie** [comment]
> So, my 2 cents (didn't look at the full discussion, sorry) is that we could tie an user to a group of repos, like:
> 
> `users` -> `git_repositories` and `git_repositories` could have for example the kind (GitHub, Gitlab, Gitea at least) maybe URL (for self-hosted instead of cloud repos) and then we could integrate with respective git providers to have tokens (that way we won't have Auth information on our side, which is a big plus for me).
> 
> As of preventing SPAM, there are some tutorials on using SpamAssassin from other projects, so maybe we can check for packages with READMEs and other docs that have high points and alert something, so we can approve or reject them?

I personally really don't want to hold any auth data. And the way its currently implemented is similar to what Atom does. You go to log in with you GitHub account, and then we just store the OAuth token locally, keeping the account tied to the GitHub account. 

But use of SpamAssassin could be nice for the the package uploading aspect, which I know we have discussed before. Although may still leave the previous concern of star spam unaddressed. 

Although using the OAuth token as a primary key could seem probamatic as those keys will change over time, and automatically expire after a certain time frame if I'm not mistaken. Although I'll have to double check.

But @Digitalone1 I have not done that migration as of yet, things have been busy. But the `docs/database.md` does need to be updated to include the new tables that have been added, which you can see the structure in `scripts/database`

But I do have a script in a private repo that I should be able to add your suggested changes, for the package versions

**Comment by Digitalone1** [comment]
There should be some parameter which define the user uniquely. 

> You can change your username to another username that is not currently in use. If the username you want is not available, consider other names or unique variations.

https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-personal-account-settings/changing-your-github-username

That said, we could even set the username as primary key. The problem is, if the user changes his/her username, how can we track it to update that state into the database? Since we interact with Github to get the OAuth token, I suppose we also get something to recognize the user even if he/she changes the name.

**Comment by confused-Techie** [comment]
@Digitalone1 While I may not have given that exact flow to much thought, I'd like to think its current configuration could still work. The current configuration being, a user entered on the DB during user creation, is given a proper auto incrementing index as you suggested previously, then GitHub OAuth does its thing, and all we get back from GitHub is the [OAuth](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#response) token. From there, we query to get the user name, and probably the user gravatar, for the website, or gravatar link that is. Then we go ahead and generate the users Pulsar Token.

The Pulsar token is all thats then exposed to the user, and any queries made that require authentication would then use the Pulsar Token, which if valid grabs the valid GitHub Token belonging to the same user, and uses it to interact with GitHub.

When the GitHub Token expires or gets revoked, trying to access GitHub API's with an expired token will fail, and Pulsar will ask to re-authenticate, but the user unique ID is still unique, and the Pulsar Token remains the same.

And if they change their name, both the Pulsar Token, GitHub Token, and Unique ID are still the same.

From my initial research it seems that this could be the way that Atom was originally doing it, since they provided a separate API token, that was different from the linked to GitHub OAuth Token.

And I'd hope that an auto-incrementing number is unique enough, or if anything we could rely on another token, since it seems near impossible that we couldn't rely on any one of those. Maybe we make sure two out of three match during user name changes or OAuth Token refreshes, or lastly against a regeneration of the Pulsar Token. Just an idea.

Otherwise if you are aware of a unique enough property for the account, that we can hopefully retrieve via GitHub I'm open to it for sure. Since setting up sign in via GitHub is something new to me, but this seems like it could work

**Comment by confused-Techie** [comment]
As for the last part of your comment about getting something back to recognize them, just wanted to be clear all we get back is the OAuth token. We would have to preform other API calls if we wanted to collection something else that is truly unique.

**Comment by Digitalone1** [comment]
@confused-Techie Thanks for the explanation! Now it's more clear to me how it works. So assuming that the Pulsar token should be unique to the user and it has to be set for the user in order to be registered, it could be a candidate for the primary key, the serial or a uuid seems useless.

In the behavior you explained, there's so issue about changing username, because we might allow the user to have a custom name for Pulser different than Github and he/she can change it, but the user is still identified by the Pulsar Token and his/her votes still count as one. So the "spamming issue" would be a user creating multiple users on Github to count more votes on Pulsar packages, but this is hard to reproduce because it's also an issue related to Github and I don't think they easily allow such behavior creating spamming accounts from nowhere.

**Comment by mauricioszabo** [comment]
@Digitalone1 it's never a good idea to have any meaningful data as primary key - I suggest keeping the UUID, it does not expose any private information, and we can use as unique identifiers anywhere

Also, I disagree about SPAM. GitHub's SPAM protection is a joke, honestly, and I hope that, in the future, we will allow people to publish packages from any git repo, so the anti-spam needs to stay on our side IMHO...

**Comment by Digitalone1** [comment]
No problem, the UUID can be used as primary key and the `unique` constraint can be set on Pulsar Token column.

**Comment by Digitalone1** [comment]
@confused-Techie I'm reading `database.js`. If we call `checkSQLSetup` at the start of every database function, why is it exported outside? We could just hide it and use it only internally.

I suppose it's invoked inside to not let the caller handle the database connection. Am I right?

**Comment by confused-Techie** [comment]
> @confused-Techie I'm reading `database.js`. If we call `checkSQLSetup` at the start of every database function, why is it exported outside? We could just hide it and use it only internally.
> 
> I suppose it's invoked inside to not let the caller handle the database connection. Am I right?

It was only exported when I was unsure of exactly how we would handle it, and was thinking maybe we just want to call the function during the initial startup. But yeah as it is currently, we can remove it and keep in internal to that module with no issues. But yeah its called just so that any caller doesn't have to worry about it being up, or handling its errors. But also as it is, pretty sure nobody is handling its errors.

Now that we think about it, there may be some value, in making that function a bit better at returning errors, and we then call it during server startup, since we will always need it. Maybe even the same with the usage of Google Cloud, since thats also called before each action in `storage.js`. Then in `server.js` we expand its startup to set both of those up, and fail if they do

**Comment by Digitalone1** [comment]
Can you link me a documentation about postgres javascript management?

The syntax 

```js
sql_storage`
  SQL QUERY TEXT
`
```

is kind of new to me.

**Comment by mauricioszabo** [comment]
@Digitalone1 it's called "String literals" and it's a new-ish thing on Javascript - basically, allows you to "customize" how you "concatenate" strings or other things.

```js
const sql = (strs, ...params) => [strs.join(' ? '), ...params]
sql`SELECT * FROM packages WHERE name=${'hello'} and tag=${'demo'}`

// Returns:
//[
//  'SELECT * FROM packages WHERE name= ?  and tag= ? ',
//  'hello',
//  'demo'
//]
```

**Comment by confused-Techie** [comment]
@Digitalone1 just make sure like you'll see in the codebase currently when using them, the `await` is what actually executes the command, otherwise you need to use I think `.execute()` on the variable. But [heres](https://github.com/porsager/postgres) more info.

**Comment by confused-Techie** [comment]
Considering that we have fully migrated to an SQL environment, and that there's just a few things to fix up before a full 1.0 release, I'll close this issue.
